include/execution.hpp

```diff
@@ -2093,7 +2093,100 @@ namespace std::execution {
   /////////////////////////////////////////////////////////////////////////////
   // [execution.senders.adaptors.upon_error]
   namespace __upon_error {
+    template <class _ReceiverId, class _Fun>
+      class __receiver
+        : receiver_adaptor<__receiver<_ReceiverId, _Fun>, __t<_ReceiverId>> {
+        using _Receiver = __t<_ReceiverId>;
+        friend receiver_adaptor<__receiver, _Receiver>;
+        [[no_unique_address]] _Fun __f_;
+
+        // Customize set_error by invoking the invocable and passing the result
+        // to the base class
+        template <class _Error>
+          requires invocable<_Fun, _Error> &&
+            __callable<set_value_t, _Receiver, invoke_result_t<_Fun, _Error>>
+        void set_error(_Error&& __error) && noexcept try {
+          execution::set_value(
+              ((__receiver&&) *this).base(),
+              std::invoke((_Fun&&) __f_, (_Error&&) __error));
+        } catch(...) {
+          execution::set_error(
+              ((__receiver&&) *this).base(),
+              current_exception());
+        }
+        
+        // Handle the case when the invocable returns void
+        template <class _R2 = _Receiver, class _Error>
+          requires invocable<_Fun, _Error> &&
+            same_as<void, invoke_result_t<_Fun, _Error>> &&
+              __callable<set_value_t, _R2>
+        void set_error(_Error&& __error) && noexcept try {
+          invoke((_Fun&&) __f_, (_Error&&) __error);
+          execution::set_value(((__receiver&&) *this).base());
+        } catch(...) {
+          execution::set_error(
+              ((__receiver&&) *this).base(),
+              current_exception());
+        }
+
+       public:
+        explicit __receiver(_Receiver __rcvr, _Fun __fun)
+          : receiver_adaptor<__receiver, _Receiver>((_Receiver&&) __rcvr)
+          , __f_((_Fun&&) __fun)
+        {}
+      };
+
+    template <class _SenderId, class _Fun>
+      class __sender
+        : sender_adaptor<__sender<_SenderId, _Fun>, __t<_SenderId>> {
```

Comment:
Hrm. On the `main` branch, the `then` sender (far above) does _not_ inherit from `sender_adaptor`, so it looks like you've haven't based this work on the most recent commit. I changed `then` sender's `connect` customization from a member function to `tag_invoke` so as to be sensitive to the cv- and ref-qualification of the sender. You'll need to make similar changes to the `upon_error` and `upon_stopped` senders.

See the changes I made in 1952036 for reference.

Reply:
From a user perspective this is bad. We should be able to reuse `sender_adaptor` here and everywhere else too.

Given that this is about qualification of the sender, I have some high hopes that `deduce this` will give us a way out of this.

No change requested

Reply:
Yes, "deduced `this`" makes the adaptors much more useful.